spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/SpannerTemplate.java

```diff
@@ -219,87 +223,116 @@ public long executeDmlStatement(Statement statement) {
 	}
 
 	@Override
 	public void upsert(Object object, String... includeProperties) {
-		applyMutations(
-				this.mutationFactory.upsert(object,
-						(includeProperties.length == 0) ? null
-								: new HashSet<>(Arrays.asList(includeProperties))));
+		Set<String> incl = (includeProperties.length == 0) ? null : new HashSet<>(Arrays.asList(includeProperties));
+		List<Mutation> mutations = this.mutationFactory.upsert(object, incl);
+		emitSaveEventsApplyMutations(mutations, Collections.singletonList(object), incl);
 	}
 
 	@Override
 	public void upsert(Object object, Set<String> includeProperties) {
-		applyMutations(this.mutationFactory.upsert(object, includeProperties));
+		List<Mutation> mutations = this.mutationFactory.upsert(object, includeProperties);
+		emitSaveEventsApplyMutations(mutations, Collections.singletonList(object), includeProperties);
+	}
+
+	private void emitSaveEventsApplyMutations(List<Mutation> mutations, Iterable entities,
+			Set<String> includeProperties) {
+		maybeEmitEvent(new BeforeSaveEvent(mutations, entities, includeProperties));
+		applyMutations(mutations);
+		maybeEmitEvent(new AfterSaveEvent(mutations, entities, includeProperties));
 	}
 
 	@Override
 	public void delete(Object entity) {
-		applyMutations(Collections.singletonList(this.mutationFactory.delete(entity)));
+		List<Mutation> mutations = Collections.singletonList(this.mutationFactory.delete(entity));
+		Iterable entities = Collections.singletonList(entity);
+		emitDeleteEventsApplyMutations(entities, mutations);
 	}
 
 	@Override
 	public void deleteAll(Iterable objects) {
-		applyMutations(
-				(Collection<Mutation>) StreamSupport.stream(objects.spliterator(), false)
-						.map(this.mutationFactory::delete).collect(Collectors.toList()));
+		List<Mutation> mutations = (List<Mutation>) StreamSupport.stream(objects.spliterator(), false)
+				.map(this.mutationFactory::delete).collect(Collectors.toList());
+		emitDeleteEventsApplyMutations(objects, mutations);
+	}
+
+	private void emitDeleteEventsApplyMutations(Iterable objects, List<Mutation> mutations) {
+		maybeEmitEvent(new BeforeDeleteEvent(mutations, objects, null, null));
+		applyMutations(mutations);
+		maybeEmitEvent(new AfterDeleteEvent(mutations, objects, null, null));
 	}
 
 	@Override
 	public void delete(Class entityClass, Key key) {
-		applyMutations(
-				Collections.singletonList(this.mutationFactory.delete(entityClass, key)));
+		List<Mutation> mutations = Collections.singletonList(this.mutationFactory.delete(entityClass, key));
+		KeySet keys = KeySet.newBuilder().addKey(key).build();
+		emitDeleteEventsApplyMutations(entityClass, keys, mutations);
 	}
 
 	@Override
 	public void delete(Class entityClass, KeySet keys) {
-		applyMutations(Collections
-				.singletonList(this.mutationFactory.delete(entityClass, keys)));
+		List<Mutation> mutations = Collections
+				.singletonList(this.mutationFactory.delete(entityClass, keys));
```

Comment:
maybe make `mutationFactory.delete` return list so you don't have to manually wrap it?